packages/shared-ui/src/utils/signin-adapter.ts

```diff
@@ -116,11 +116,19 @@ class SigninAdapter {
    * signed out.
    */
   async token(): Promise<ValidTokenResult | SignedOutTokenResult> {
-    const token = this.#tokenVendor.getToken(SIGN_IN_CONNECTION_ID);
+    let token = this.#tokenVendor.getToken(SIGN_IN_CONNECTION_ID);
     if (token.state === "expired") {
-      return token.refresh();
+      token = await token.refresh();
+      if (token.state === "signedout") {
+        if ((await this.signIn()).ok) {
+          token = this.#tokenVendor.getToken(SIGN_IN_CONNECTION_ID);
+        }
+      }
+    }
+    if (token.state === "valid" || token.state === "signedout") {
+      return token;
     }
-    return token;
+    throw new Error("Invalid token state after refresh: " + token.state);
```

Comment:
can this be thrown inside the `if (token.state === "expired")` block? imagine we added a new state -- we wouldn't want to use this error for that case.